### PR TITLE
Finished the first version of assignment3, will add more javadoc comm…

### PR DESCRIPTION
**Pull request by Adelina9**
…ents later

### CODE REVIEW

File: Adelina9/assignment-3/Trie.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            }
+        }
+    }
+
+    // combining an x and y value we get one of the possible movements in the grid
+    static int x[] = { -1, 0, 1 };
+    static int y[] = { -1, 0, 1 };
+
+    // function that returns the words found in the grid
+    public List<String> findWordsGrid(char[][] grid, int[][] visited, ArrayList<String> words) {
```
Comment (TDiva) [inline Adelina9/assignment-3/Trie.java]: Maybe we better export this function to some utility class? This is not related to trie functionality.

File: Adelina9/assignment-3/Trie.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                if (mutableBoolean.getValue())
+                    return true;
+                else
+                    visited[i][j] = 0;
+            }
+        }
+        return false;
+    }
+
+    // function that traverses the grid recursively and searches an word
```
Comment (TDiva) [inline Adelina9/assignment-3/Trie.java]: s/an/the

File: Adelina9/assignment-3/Trie.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    private void searchGrid(int i, int j, String word, int index, MutableBoolean mutableBoolean, char[][] grid,
```
Comment (TDiva) [inline Adelina9/assignment-3/Trie.java]: This also belongs to some other class I believe.

Comment (TDiva) [reply Adelina9/assignment-3/Trie.java]: \+ why not return this boolean instead of using mutable one?

File: Adelina9/assignment-3/TrieNode.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
+/**
+ * TrieNode class that stores if a node is end of word and an array of children
+ * for the node.
+ */
+public class TrieNode {
+    // number of letters in the english alphabet
+    static final int LETTERS_ENGLISH = 26;
+
+    private TrieNode[] children = new TrieNode[LETTERS_ENGLISH];
```
Comment (TDiva) [inline Adelina9/assignment-3/TrieNode.java]: Finally I've got to this line :)
The problem is exactly the same as with anagrams.
 * Why do we need to store whole array of unused references?
 * Why do we limit our implementation to english letters?

Use HashMap for such cases and forget about arrays with explicit indices :)

File: Adelina9/assignment-3/TestLexicon.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * Test class that tests isWord, isPrefix and findAllWords methods.
+ * 
+ */
+public class TestLexicon {
+
+    // assumed that grid is no bigger than this
+    // can be resized for bigger inputs
+    static char grid[][] = new char[2000][2000];
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/TestLexicon.java]: Done.

File: Adelina9/assignment-3/Trie.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Trie class that implements insert and search operations. Also, implements
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Done.

File: Adelina9/assignment-3/Trie.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+/**
+ * Trie class that implements insert and search operations. Also, implements
+ * method that find all the words from the lexicon in a given grid.
+ */
+public class Trie {
+    public Trie() {
+    }
+
+    static final int LETTERS_ENGLISH = 26;
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Changed the name of the constant to LETTERS to make it more general and added a comment where I said about the assumption that only english characters are used.

File: Adelina9/assignment-3/Trie.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    private TrieNode root;
+    private int wordNumber;
+
+    public int getWordNumber() {
+        return this.wordNumber;
+    }
+
+    public void setWordNumber(int wordNumber) {
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Made the setter and getter for wordNumber private.

File: Adelina9/assignment-3/Trie.java [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        return root;
+    }
+
+    public void setRoot(TrieNode root) {
+        this.root = root;
+    }
+
+    // if not present inserts key
+    // if key is prefix mark leaf node
+    void insert(String key) {
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Checked if the characters of the key consist of english letters only, else throw InputMismatchException. Added a test for this case.

File: Adelina9/assignment-3/Trie.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        int length = key.length();
+        int index;
+
+        TrieNode pCrawl = root;
+
+        for (level = 0; level < length; level++) {
+            index = key.charAt(level) - 'a';
+            if (pCrawl.getChildren()[index] == null) {
+                pCrawl.getChildren()[index] = new TrieNode();
+                setWordNumber(getWordNumber() + 1);
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Done.

File: Adelina9/assignment-3/Trie.java [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        for (level = 0; level < key.length(); level++) {
+            index = key.charAt(level) - 'a';
+
+            if (pCrawl.getChildren()[index] == null)
+                return false;
+
+            pCrawl = pCrawl.getChildren()[index];
+        }
+
+        return (pCrawl != null);
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Made a private function that checks all the letters and has a parameter that indicates if the last letter in the key is end of word.

File: Adelina9/assignment-3/Trie.java [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    // function that returns a list containing all the words int the trie
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: I'm not sure what that means.

File: Adelina9/assignment-3/Trie.java [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    public ArrayList<String> getWordsTrie(char str[], int level) {
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: I meant get words FROM trie, but it looks like I made it less sugestive this way. Removed the word trie, now it's called getWords.

File: Adelina9/assignment-3/Trie.java [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+        ArrayList<String> words = new ArrayList<>();
+        findWordsTrie(root, str, level, words);
+        return words;
+    }
+
+    private boolean isLeafNode(TrieNode node) {
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Removed.

File: Adelina9/assignment-3/Trie.java [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+        List<String> foundWords = new ArrayList<>();
+
+        for (int i = 0; i < words.size(); i++)
+            if (findWordGrid(words.get(i), grid, visited))
+                foundWords.add(words.get(i));
+
+        return foundWords;
+    }
+
+    // function that searches for a single word inf the grid
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Done.

File: Adelina9/assignment-3/Trie.java (hunk 1 above)
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Done.

File: Adelina9/assignment-3/Trie.java (hunk 2 above)
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Done.

File: Adelina9/assignment-3/Trie.java (hunk 3 above)
Comment (Adelina9) [reply Adelina9/assignment-3/Trie.java]: Now returns the boolean insead of using the mutable one. Moved the function to the grid utiliy class.

File: Adelina9/assignment-3/MutableBoolean.java [hunk 16]
```diff
@@ -0,0 +1,15 @@
+class MutableBoolean {
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-3/MutableBoolean.java]: Not using the class anymore.

File: Adelina9/assignment-3/TrieNode.java (hunk 4 above)
Comment (Adelina9) [reply Adelina9/assignment-3/TrieNode.java]: Done :)

File: Adelina9/assignment-3/Grid.java [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    private static boolean searchGrid(int i, int j, String word, List<String> words, char[][] grid, int[][] visited,
+            Trie lexicon) {
+        // word is also a prefix so do not mark the cell as visited
+        if (lexicon.isWord(word)) {
+            words.add(word);
+            return true;
+        }
+        for (int a = 0; a < 3; a++) {
+            if (i + x[a] > -1 && i + x[a] < grid.length) {
```
Comment (TDiva) [inline Adelina9/assignment-3/Grid.java]: What about creating a function which checks if a cell is inside grid borders? Then you can remove all these nested calls which are quite hard to read.
```
for (int dx: x) {
  for (int dy: y) {
    if (!checkInBorders(i + dx, j + dy)) { continue; }
    . . .
  }
}
```

Also, try to avoid indices like `i` and `j` outside of simple loops, better use full names. It's easy to get confused what is row and what is column.

File: Adelina9/assignment-3/Grid.java [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+     * Space complexity: O(8^(n * m)), n - rows in the grid, m - columns in the grid, from the recursion
+     * 
+     * @param grid    the grid in which we search the words
+     * @param visited two-dimensional array to mark the visited cells
+     * @param lexicon the lexicon that contains the words to search
+     * @return the words found in the grid
+     */
+    public static List<String> findWordsGrid(char[][] grid, int[][] visited, Trie lexicon) {
+        List<String> foundWords = new ArrayList<>();
+        Grid.searchGrid(0, 0, "", foundWords, grid, visited, lexicon);
```
Comment (TDiva) [inline Adelina9/assignment-3/Grid.java]: Doesn't this check only words starting at cell [0,0]?

Comment (Adelina9) [reply Adelina9/assignment-3/Grid.java]: No, it's just the starting point.

File: Adelina9/assignment-3/Grid.java (hunk 17 above)
Comment (Adelina9) [reply Adelina9/assignment-3/Grid.java]: Done.

File: Adelina9/assignment-3/Grid.java [hunk 19]
```diff
@@ -12,63 +12,72 @@
...
-    public static List<String> findWordsGrid(char[][] grid, int[][] visited, Trie lexicon) {
-        List<String> foundWords = new ArrayList<>();
-        Grid.searchGrid(0, 0, "", foundWords, grid, visited, lexicon);
+    public static Set<String> findWordsGrid(char[][] grid, int[][] visited, Trie lexicon) {
+        Set<String> foundWords = new HashSet<String>();
+        for (int i = 0; i < grid.length; i++) {
+            for (int j = 0; j < grid[0].length; j++) {
+                String initWord = Character.toString(grid[i][j]);
+                if (lexicon.isPrefix(initWord)) {
+                    visited[i][j] = 1;
```
Comment (TDiva) [inline Adelina9/assignment-3/Grid.java]: No need to pass `visited` sa parameter - the exported function should have only grid and lexicon.

File: Adelina9/assignment-3/Grid.java [hunk 20]
```diff
@@ -12,63 +12,72 @@
...
      */
-    private static boolean searchGrid(int i, int j, String word, List<String> words, char[][] grid, int[][] visited,
-            Trie lexicon) {
-        // word is also a prefix so do not mark the cell as visited
+    private static void searchGrid(int row, int column, String word, Set<String> words, char[][] grid, int[][] visited,
+           Trie lexicon) {
         if (lexicon.isWord(word)) {
-            words.add(word);
-            return true;
+            if (!words.contains(word))
```
Comment (TDiva) [inline Adelina9/assignment-3/Grid.java]: `words.add(word)` for Set will ensure there're no duplicates, no need to check it twice.

File: Adelina9/assignment-3/Grid.java [hunk 21]
```diff
@@ -12,63 +12,72 @@
...
+                words.add(word);
+            if (!lexicon.isPrefix(word))
```
Comment (TDiva) [inline Adelina9/assignment-3/Grid.java]: I think this check should be outside of the `if (lexicon.isWord(word)) { ... }` check)

File: Adelina9/assignment-3/Grid.java [hunk 22]
```diff
@@ -12,63 +12,72 @@
...
-                                } else if (lexicon.isWord(newWord)) {
-                                    words.add(newWord);
-                                    visited[i + x[a]][j + y[b]] = 1;
-                                }
-                            }
-                        }
-                    }
+        for (int dx : x) {
+            for (int dy : y) {
+                if (Grid.checkInBorders(row + dy, column + dx, grid.length, grid[0].length) && (dx != 0 || dy != 0)
```
Comment (TDiva) [inline Adelina9/assignment-3/Grid.java]: you use `row + dy` and `column + dx` very often, it's easy to get mistaken - why not create a special variable, like `neighbour_row = row = dy`?

File: Adelina9/assignment-3/Grid.java [hunk 23]
```diff
@@ -12,63 +12,72 @@
...
+                        && visited[row + dy][column + dx] == 0) {
+                    String newWord = word.concat(String.valueOf(grid[row + dy][column + dx]));
+                    if (lexicon.isPrefix(newWord)) {
```
Comment (TDiva) [inline Adelina9/assignment-3/Grid.java]: This is duplicate check (see code above), maybe we can avoid it and check `isPrefix` only once?

File: Adelina9/assignment-3/Grid.java [hunk 24]
```diff
@@ -12,63 +12,72 @@
...
+                            words.add(newWord);
                 }
             }
         }
-        return true;
+        return;
+    }
+
+    private static boolean checkInBorders(int row, int column, int rowLimit, int columnLimit) {
+        if (row > -1 && row < rowLimit && column > -1 && column < columnLimit)
```
Comment (TDiva) [inline Adelina9/assignment-3/Grid.java]: this can be replaced with
```
  return (row > -1 && row < rowLimit && column > -1 && column < columnLimit);
```

File: Adelina9/assignment-3/Tests.java [hunk 25]
```diff
@@ -99,8 +99,64 @@ public void testFindWordsGrid() {
...
+        expectedWords.add("card");
+        
+        assertEquals(foundWords, expectedWords);
+    }
+
+    @Test
+    public void testFindWordsGrid2() {
+        char grid[][] = new char[2][3];
+        int visited[][] = new int[2][3];
+        grid[0][0] = 'a';
```
Comment (TDiva) [inline Adelina9/assignment-3/Tests.java]: maybe you could export grid and lexicon creation to some builder method? or make it part of test data?

**Comment by TDiva** [comment]
Almost done!
There are only small nits to perfect code ;)

File: Adelina9/assignment-3/Grid.java (hunk 20 above)
Comment (Adelina9) [reply Adelina9/assignment-3/Grid.java]: Done.

File: Adelina9/assignment-3/Grid.java (hunk 21 above)
Comment (Adelina9) [reply Adelina9/assignment-3/Grid.java]: Done.

File: Adelina9/assignment-3/Grid.java (hunk 22 above)
Comment (Adelina9) [reply Adelina9/assignment-3/Grid.java]: Done.

File: Adelina9/assignment-3/Grid.java (hunk 24 above)
Comment (Adelina9) [reply Adelina9/assignment-3/Grid.java]: Done.

File: Adelina9/assignment-3/Grid.java (hunk 19 above)
Comment (Adelina9) [reply Adelina9/assignment-3/Grid.java]: Done.

File: Adelina9/assignment-3/Tests.java (hunk 25 above)
Comment (Adelina9) [reply Adelina9/assignment-3/Tests.java]: Added parameterized tests for findWordsGrid in separate class TestGrid.

File: Adelina9/assignment-3/GridTest.java [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+    private Set<String> expectedWords;
+
+    public GridTest(Trie inputLexicon, char[][] inputGrid, Set<String> expectedWords) {
+        this.inputLexicon = inputLexicon;
+        this.inputGrid = inputGrid;
+        this.expectedWords = expectedWords;
+    }
+
+    @Test
+    public void test() {
```
Comment (TDiva) [inline Adelina9/assignment-3/GridTest.java]: Yep, that's very short and clear definition of tests, but it's not so nice due to no "labels" for test. The general testing approach is not to run code on a bunch of inputs, but to cover it's behavior well and be able to understand quickly why smth failed.

That's why having test cases like "emptyInput", "TooBigWord" and so on are very important to check, and parametrized tests are helping with easier specification of individual test case.

Comment (Adelina9) [reply Adelina9/assignment-3/GridTest.java]: Added some more specific tests for the findWordsGrid method.
